### PR TITLE
stop linting [`blocks_in_conditions`] on `match` with weird attr macro case

### PR DESCRIPTION
**Pull request by J-ZhengLi**
should fixes: #12016 

---

changelog: [`blocks_in_conditions`] - fix FP on `match` with weird attr macro

This might not be the best solution, as the root cause (i think?) is the `span` of block was incorrectly given by the compiler?

I'm open to better solutions

### CODE REVIEW

**Comment by J-ZhengLi** [comment]
Someone help me look at this plz?

@rustbot r? @rust-lang/clippy

File: clippy_lints/src/blocks_in_conditions.rs [hunk 1]
```diff
@@ -135,3 +134,21 @@ impl<'tcx> LateLintPass<'tcx> for BlocksInConditions {
...
+    let block_first_stmt_span = block.stmts.first().map(|st| st.span);
+    block_span.from_expansion()
+        || matches!(block_expr_span, Some(sp) if sp.from_expansion())
+        || matches!(block_first_stmt_span, Some(sp) if sp.from_expansion())
+        || {
+            // HACK: When expr and stmt overlaps, assume it was modified by some macro.
+            if let Some(expr_span) = block_expr_span
+                && let Some(stmt_span) = block_first_stmt_span
+            {
+                expr_span.overlaps(stmt_span)
```
[the remark this answers is not in the document]
Comment (J-ZhengLi) [reply clippy_lints/src/blocks_in_conditions.rs:149]: > This looks like it's sort of trying to do what [`is_from_proc_macro`](https://doc.rust-lang.org/nightly/nightly-rustc/clippy_utils/fn.is_from_proc_macro.html) does. Can we use that function, with the condition/scrutinee expression as an argument?

I tried it at first but it didn't work unfortunately...

I think it's because the macro is using 'span_quote', making the span pass to it being treated as an expansion target? And then 'is_from_proc_macro' wasn't able to give a correct result, as it only sees the 'span' usd in 'span_quote'...?

Ahh. I don't know how to explain it clearly so here is an example:

```rust
#[foo]
match x().await {
    // ...
}

in short, this code triggers the warning because `foo` expands the match scrutine to something wrapped with block, but I think the compiler wasn't able to identify the block origin? Because when getting the span of the block, it's stmts, and it's expr, they all pointing to the span of `await`, which is exactly the span used in an `span_quote` call inside of `foo` macro...

This seems like a rustc problem tho, if it somehow got fixed then I can just close this, as the original code already checks for `is_from_macro`, I might've open this PR hoping to get a quick fix lol 😆

[the remark this answers is not in the document]
Comment (J-ZhengLi) [reply clippy_lints/src/blocks_in_conditions.rs:154]: sadly it didn't work as the same, 

It looks like because of the span of `await` was used in `quote_spanned`, causing the span of anything related to the block (which was inside macro expansion) pointing back to `await`.

Therefore both `from_expansion` and `is_from_proc_macro` did not tell the susiness, since they only saw the span of `await`? That's my wild guess tho...

[the remark this answers is not in the document]
Comment (y21) [reply clippy_lints/src/blocks_in_conditions.rs:154]: > It looks like because of the span of await was used in quote_spanned, causing the span of anything related to the block (which was inside macro expansion) pointing back to await.

That's surprising, `is_from_proc_macro` should be able to deal with exactly this.
It checks if a HIR node matches roughly with what the source string at the span is. In this case, the span is `await`, but the node is a block (starts with `{`), so the mismatch should tell it that this is from a proc macro.

I just tried it real quick myself and this seems to work fine though?
<details>
<summary>Diff</summary>

```diff
diff --git a/clippy_lints/src/blocks_in_conditions.rs b/clippy_lints/src/blocks_in_conditions.rs
index 4108ff72a..fd2190683 100644
--- a/clippy_lints/src/blocks_in_conditions.rs
+++ b/clippy_lints/src/blocks_in_conditions.rs
@@ -88,7 +88,7 @@ impl<'tcx> LateLintPass<'tcx> for BlocksInConditions {
                         );
                     }
                 } else {
-                    if block_is_from_expansion(block, expr.span) {
+                    if block_is_from_expansion(block, expr.span) || clippy_utils::is_from_proc_macro(cx, cond) {
                         return;
                     }
                     // move block higher
@@ -141,14 +141,4 @@ fn block_is_from_expansion(block: &Block<'_>, block_span: Span) -> bool {
     block_span.from_expansion()
         || matches!(block_expr_span, Some(sp) if sp.from_expansion())
         || matches!(block_first_stmt_span, Some(sp) if sp.from_expansion())
-        || {
-            // HACK: When expr and stmt overlaps, assume it was modified by some macro.
-            if let Some(expr_span) = block_expr_span
-                && let Some(stmt_span) = block_first_stmt_span
-            {
-                expr_span.overlaps(stmt_span)
-            } else {
-                false
-            }
-        }
 }
```
</details>

[the remark this answers is not in the document]
Comment (J-ZhengLi) [reply clippy_lints/src/blocks_in_conditions.rs]: yes, I was just being dumb :man_facepalming: , I pass the whole `expr` into it, no wonder it's not working...

It should be good now, thanks for your advise!

**Review by y21 — APPROVED** [review]

**Comment by y21** [comment]
Thank you!
@bors r+
